feat: add nix flake support

lessuselesss opened the pull request:
# Relates to:
[Issue #1142](https://github.com/ai16z/eliza/issues/1142)

# Risks
Low - This change:
- Only affects development environment setup
- Doesn't modify runtime code
- Is optional (developers can still use traditional npm/pnpm setup)
- Can be easily reverted if issues arise

# Background

## What does this PR do?
Adds Nix Flake support to provide a reproducible development environment with:
- Correct Node.js and pnpm versions
- Helpful welcome message showing common commands
- Integration with existing monorepo structure

## What kind of change is this?
Improvements (adds optional development tooling without changing existing functionality)

# Documentation changes needed?
My changes require a change to the project documentation.
I will update the local development guide to include:
1. Installation of Nix using [Determinate Nix Installer](https://github.com/DeterminateSystems/nix-installer)
2. Instructions for using the development environment

# Testing

## Where should a reviewer start?
1. Install Nix using Determinate Nix Installer:
```bash
curl --proto '=https' --tlsv1.2 -sSf -L https://install.determinate.systems/nix | sh -s -- install
```

2. Clone the PR and enter the development environment:
```bash
git clone https://github.com/ai16z/eliza.git
cd eliza
nix develop
```

3. Verify the welcome message appears with instructions for:
   - pnpm i
   - pnpm build
   - pnpm clean

## Detailed testing steps
1. Prerequisites:
   - Install Nix following the steps above
   - Verify flakes are enabled by default

2. Test environment setup:
   ```bash
   git clone https://github.com/ai16z/eliza.git
   cd eliza
   nix develop
   ```
   - Verify welcome message appears
   - Verify Node.js version matches project requirements
   - Verify pnpm is available

3. Test build process:
   ```bash
   pnpm i
   pnpm build
   ```
   - Verify all dependencies install correctly
   - Verify build completes successfully

4. Test clean process:
   ```bash
   pnpm clean
   pnpm i
   pnpm build
   ```
   - Verify clean removes build artifacts
   - Verify rebuild works after clean

## Discord username
Adam Turner | lessuseless
ar4s_45979

shakkernerd committed:
Merge pull request #1080 from ai16z/fix/plugin-near

chore: add npmignore file

shakkernerd committed:
chore: fix broken pnpm lockfile

shakkernerd committed:
Merge pull request #1081 from ai16z/fix/pnpm-lockfile

chore: fix broken pnpm lockfile

shakkernerd committed:
fix: eslint command

shakkernerd committed:
Merge pull request #1082 from ai16z/fix/client-slack

fix: eslint command

shakkernerd committed:
chore: remove unnecessary packages

shakkernerd committed:
Merge pull request #1083 from ai16z/fix/client-slack

chore: remove unnecessary packages

shakkernerd committed:
fix: rename state variable

shakkernerd committed:
fix: rename state variable

shakkernerd committed:
chore: remove unused imports

shakkernerd committed:
fix: rename state variable

shakkernerd committed:
chore: remove unused imports

shakkernerd committed:
fix: dynamic import of fs module

shakkernerd committed:
Merge pull request #1084 from ai16z/fix/client-slack

fix: dynamic import of fs module

shakkernerd committed:
chore: remove unused imports and rename runtime variable

shakkernerd committed:
Merge pull request #1085 from ai16z/fix/client-slack

chore: remove unused imports and rename runtime variable

shakkernerd committed:
chore: delete unnecessary file

shakkernerd committed:
fix: remove unused import variables

shakkernerd committed:
fix: log error to the console

shakkernerd committed:
chore: pnpm lockfile

shakkernerd committed:
Merge pull request #1086 from ai16z/fix/client-slack

fix: client slack linting errors

shakkernerd committed:
fix: syntax error: invalid arithmetic operator

shakkernerd committed:
fix: invalid arithmetic operator

oxSaturn committed:
feat: allow users to configure models for grok

shakkernerd committed:
Merge pull request #1088 from ai16z/fix/smokeTests

fix: syntax error: invalid arithmetic operator

shakkernerd committed:
chore: increase timeout to 3mins

shakkernerd committed:
Merge pull request #1092 from ai16z/fix/smokeTests

chore: increase timeout to 3mins

shakkernerd committed:
chore: kill pnpm start

shakkernerd committed:
Merge pull request #1093 from ai16z/fix/smokeTests

chore: kill pnpm start

shakkernerd committed:
chore: debugging start behaviour

nicky-ru committed:
docs: add readme to plugin-evm

shakkernerd committed:
Merge pull request #1094 from ai16z/fix/smokeTests

chore: debugging start behaviour

imthatcarlos committed:
Merge branch 'develop' into feat/client-lens

BlockJuic3 committed:
fix: return types of createAgent & startAgent

shakkernerd committed:
Merge pull request #1095 from nicky-ru/docs-plugin-evm-readme

docs: add README.md to plugin-evm

xwxtwd committed:
Merge branch 'develop' into feat/add_plugin-nft-generator

# Conflicts:
#	agent/src/index.ts
#	packages/core/src/types.ts
#	pnpm-lock.yaml

azep-ninja committed:
Merge branch 'develop' into discord-telegram-team-agents-and-optimizations

mcp97 committed:
Merge branch 'develop' into fix/agent-functions-return-typ

mcp97 committed:
Merge pull request #1097 from BlockJuic3/fix/agent-functions-return-typ

fix: return types of createAgent & startAgent

mcp97 committed:
Merge branch 'main' into main

madjin committed:
add bankless and twitter space notes

shakkernerd committed:
feat: Smoke Test script

shakkernerd committed:
Merge pull request #1101 from ai16z/fix/smokeTests

feat: Smoke Test script

mcp97 committed:
Merge pull request #1091 from oxSaturn/feat/configure-models-grok-2

feat: allow users to configure models for grok

odilitime committed:
Merge pull request #998 from ai16z-demirix/main

test: adding tests. changed files actions.test.ts, messages.test.ts, models.test.ts

azep-ninja committed:
Merge branch 'develop' into discord-telegram-team-agents-and-optimizations

BalanaguYashwanth committed:
Merge branch 'develop' of https://github.com/BalanaguYashwanth/eliza into 656--fix-twitter-multi-agent-support

odilitime committed:
Merge pull request #1036 from BalanaguYashwanth/656--fix-twitter-multi-agent-support

chore: improving client typing

shakkernerd committed:
fix: handle no termination mesage

shakkernerd committed:
Merge pull request #1102 from ai16z/fix/smokeTests

fix: handle no termination message

shakkernerd committed:
chore: fix broken pnpm lockfile

shakkernerd committed:
Merge pull request #1103 from ai16z/fix/pnpm-lockfile

chore: fix broken pnpm lockfile

mcp97 committed:
Merge pull request #1070 from Titan-Node/main

fix: Fixed twitter posts include from including `/n` in the text

shakkernerd committed:
fix: add missing imports

mcp97 committed:
Merge pull request #1104 from ai16z/fix/starting_agent

fix: add missing imports

lalalune committed:
Merge branch 'main' into patch-1

lalalune committed:
Merge branch 'develop' into patch-1

lalalune committed:
Merge pull request #897 from guzus/patch-1

chore: improve formatting of .env.example for better readability

lalalune committed:
fix missing

lalalune committed:
Merge branch 'develop' of http://github.com/ai16z/eliza into develop

lalalune committed:
Merge branch 'feat/client-lens' of https://github.com/mad-finance/eliza into develop

lalalune committed:
Merge pull request #1011 from xwxtwd/feat/add_plugin-nft-generator

feat: Add plugin-nft-generation: create Solana NFT collections.

brunocalmels committed:
Fix title levels in plugins docs

lalalune committed:
commit lock

brunocalmels committed:
Fix typo

lalalune committed:
Merge pull request #1107 from brunocalmels/patch-1

Documentation: Plugin list numbering and titles

lalalune committed:
Merge branch 'jnaulty/add-sui-plugin' of https://github.com/jnaulty/eliza into develop

lalalune committed:
Merge branch 'develop' of http://github.com/ai16z/eliza into develop

lalalune committed:
Merge branch 'plugin-story' of https://github.com/storyprotocol/eliza into develop

lalalune committed:
fix build

lalalune committed:
Merge branch 'fix/charity' of http://github.com/ai16z/eliza into develop

madjin committed:
update community docs

madjin committed:
merge with develop

mcp97 committed:
Merge pull request #1111 from ai16z/new_docs

Feat: Update community section of docs

mcp97 committed:
Revert "Feat: Update community section of docs"

madjin committed:
Merge pull request #1112 from ai16z/revert-1111-new_docs

fix: Revert "Feat: Update community section of docs"

madjin committed:
update docs

madjin committed:
update api docs, add dotenv module

madjin committed:
add profiles page

madjin committed:
Merge branch 'develop' of github.com:ai16z/eliza into new_docs

madjin committed:
update docs

mcp97 committed:
Merge pull request #1114 from ai16z/new_docs

feat: New docs for community section

shakkernerd committed:
chore: fix broken pnpm lockfile

shakkernerd committed:
Merge pull request #1115 from ai16z/fix/pnpm-lockfile

chore: fix broken pnpm lockfile

lalalune committed:
Merge pull request #837 from qgpcybs/main

fix: load image from diff endpoints

lalalune committed:
use large models

lalalune committed:
Merge branch 'develop' of http://github.com/ai16z/eliza into HEAD

lalalune committed:
Merge pull request #957 from azep-ninja/discord-telegram-team-agents-and-optimizations

feat: Updated characters types, Discord & Telegram enhancements

lalalune committed:
Merge branch 'develop' into 813-disable-fal-safety-checker

lalalune committed:
Merge pull request #814 from daojonesceo/813-disable-fal-safety-checker

FAL image settings escape hatch

mcp97 committed:
fix: prevent pnpm lock file from incorrectly being updated on install and add explicit update script to update use w/o frozen lock file option

mcp97 committed:
remove -r

mcp97 committed:
on default don't update pnpm lockfile

mcp97 committed:
remove space

mcp97 committed:
don't need to explicitly set it with .npmrc in root file

mcp97 committed:
remove space

mcp97 committed:
Update quickstart.md

include the scenarios to not freeze the lockfile

0xCardiE committed:
add 2 new chars

0xCardiE committed:
remove adding new chars

0xCardiE committed:
add missing params

0xCardiE committed:
give proper voice

azep-ninja committed:
discord fix - duplicate function

lalalune committed:
Merge pull request #1124 from n00b21337/add_example_chars

fix: add more heplful default agents (Dobby and 3CPO)

mcp97 committed:
Merge pull request #1125 from azep-ninja/fix/discord-duplicats-fix

fix: discord client duplicate function removal

mcp97 committed:
Merge branch 'develop' into realitySpiral/lockFileFix

mcp97 committed:
Merge pull request #1120 from Sifchain/realitySpiral/lockFileFix

fix: Refactor to prevent unnecessary lockfile changes

MbBrainz committed:
add chatapi by akash network to model configs

lalalune committed:
update banner

lalalune committed:
Merge branch 'main' of http://github.com/ai16z/eliza into develop

0xCardiE committed:
fix name

0xCardiE committed:
add new file name

jzvikart committed:
Increase delay to avoid CI/CD timeouts

jzvikart committed:
Refactoring

jzvikart committed:
Wait for server to start

jzvikart committed:
Fixes

jzvikart committed:
Cleanup

MbBrainz committed:
Merge branch 'develop' into main

mcp97 committed:
Merge pull request #1133 from n00b21337/fix/add_example_chars

fix: fix the name

mcp97 committed:
Merge pull request #1131 from MbBrainz/main

Add `chatapi.akash.network` to available list of model providers (FREE LLAMA API ACCESS!)

Commit:
added FOMO createToken action in plugin-solana

Commit:
feat: add support for handlebars templating engine as an option

Commit:
feat: document advanced usage in md file

Commit:
feat: after CR, removing long example

mcp97 committed:
Merge pull request #1136 from erise133/feat/handlebars-templating-engine

feat: add support for handlebars templating engine as an option

jzvikart committed:
Get stopAgent() to work as expected

jzvikart committed:
Cleanup

jzvikart committed:
Avoid using console.clear() to preserve messages for debugging

rarepepi committed:
remove docker compose command since Docker file already runs

azep-ninja committed:
fix tg duplicate functions.

owlcode committed:
clean newlines for new tweet

mcp97 committed:
Merge pull request #1141 from owlcode/clean-newlines-from-new-tweets

clean newlines for new tweet

madjin committed:
Merge branch 'develop' of github.com:ai16z/eliza into new_docs

odilitime committed:
Merge branch 'develop' into fix/plugin-evm

mcp97 committed:
Merge pull request #1140 from azep-ninja/fix/duplicate-tg-funtions

fix: telegram client duplicate function removal

odilitime committed:
fix merge: remove double improve, adjust params to various calls, use initWalletProvider

odilitime committed:
Merge pull request #965 from FWangZil/fix/plugin-evm

fix: Fix Parameter Parsing in plugin-evm TransferAction and Return Transaction Hash

odilitime committed:
Merge pull request #1135 from fomoTon/fomo-token-plugin

feat: allow agents to create/buy/sell tokens on FOMO.fund's bonding curve in plugin-solana

odilitime committed:
update lockfile for PR1135

shakkernerd committed:
chore: remove comment

shakkernerd committed:
Merge pull request #1143 from ai16z/remove-comment

chore: remove comment

shakkernerd committed:
fix: remove unused import

shakkernerd committed:
fix: Expected an assignment or function call and instead saw an expression

shakkernerd committed:
fix: use of let instead of const

shakkernerd committed:
chore: comment out unused imports

shakkernerd committed:
fix: remove unused variable

shakkernerd committed:
chore: lint command

shakkernerd committed:
chore: add eslint config

shakkernerd committed:
chore: delete client-whatsapp folder

mcp97 committed:
Merge pull request #1139 from rarepepi/docker-fixes

fix: remove docker compose command since Docker file already runs

odilitime committed:
improve on fomo plugin and distingush it from pump.fun's plugin

shakkernerd committed:
Merge pull request #1147 from odilitime/fix-lint

fix: improve fomo integration

odilitime committed:
include fomo

odilitime committed:
update URLs and example response

odilitime committed:
Merge branch 'develop' of https://github.com/ai16z/eliza into fix-lint

shakkernerd committed:
Merge pull request #1148 from odilitime/fix-lint

chore: fix PR #1147

odilitime committed:
Merge pull request #1144 from ai16z/develop

chore: Merge monday, merging develop into main

actions-user committed:
chore: update changelog

shakkernerd committed:
feat: update packages version script

jzvikart committed:
Merge remote-tracking branch 'ai16z/develop' into realitySpiral/integration-tests

shakkernerd committed:
Merge pull request #1150 from ai16z/update-version

feat: update packages version script

jzvikart committed:
Remove dependency zx

shakkernerd committed:
chore: bump version to 0.1.6-alpha.3

shakkernerd committed:
Merge pull request #1152 from ai16z/new_version

chore: bump version to 0.1.6-alpha.3

jzvikart committed:
Minor cleanup

shakkernerd committed:
fix: fetch log level to debug

shakkernerd committed:
Merge pull request #1153 from ai16z/fetch_logs_debug

fix: fetch log level to debug

odilitime committed:
fix directClient

mcp97 committed:
Merge pull request #1154 from odilitime/fix-lint

fix: fix direct-client ability to start agents

shakkernerd committed:
Merge pull request #1155 from ai16z/develop

chore: develop into main

actions-user committed:
chore: update changelog

jzvikart committed:
Disable logging of fetch requests to avoid disclosing sensitive information

madjin committed:
update community docs

tcm390 committed:
allow multiple bots to join the voice channel

tcm390 committed:
allow multiple bots to join the voice channel

shakkernerd committed:
chore: fix broken pnpm lockfile

shakkernerd committed:
fix: client twitter login and auth handler

shakkernerd committed:
Merge pull request #1158 from ai16z/fix/client-twitter

fix: client twitter login and auth handler

yang-han committed:
fix: print commands to start the client and remove unused --non-iteractive in dockerfile

yang-han committed:
fix: print commands to start the client and remove unused --non-iteractive in dockerfile

shakkernerd committed:
chore: bump version to 0.1.6-alpha.4

shakkernerd committed:
Merge pull request #1159 from ai16z/new_version

chore: bump version to 0.1.6-alpha.4

actions-user committed:
chore: update changelog

HashWarlock commented:
@lessuselesss love this PR, but there are some weird problems that will cause a NixOS user to fail when building the codebase with nix flakes enabled.

For example, I built this on my NixOS machine and we see this error:
```
WARN  Unsupported engine: wanted: {"node":"23.3.0"} (current: {"node":"v20.18.1","pnpm":"9.15.0"})
docs                                     |  WARN  Unsupported engine: wanted: {"node":"23.3.0"} (current: {"node":"v20.18.1","pnpm":"9.15.0"})
```

We may think...what?! No Way...But how?? The pkgs specifically lists `nodejs_23` and when I run `node version` I will see the `v23.2.0`, but that still does not equal `v20.18.1`.

So I did some digging bc Nix can be a pain in the ass at times with weird dependencies errors. So I checked the `pnpm` pkgs source code and found this line https://github.com/NixOS/nixpkgs/blob/394571358ce82dff7411395829aa6a3aad45b907/pkgs/development/tools/pnpm/generic.nix#L28

And `nodejs` pkg points to:
![image](https://github.com/user-attachments/assets/1e258b67-924e-4471-a590-d7bde3ac7c64)

So this here is the culprit for why a NixOS user will hit this weird error even though we declaratively chose the right node version.

shakkernerd committed:
Merge pull request #1156 from ai16z/tcm-discord-multiple-bot

fix: Enable multiple bots to join Discord voice channels

mcp97 committed:
Merge pull request #1163 from yang-han/fix_client

chore: print commands to start the client and remove unused --non-itera…

shakkernerd committed:
feat: make script dash compatible

shakkernerd committed:
Merge pull request #1165 from ai16z/fix/start_script

feat: make script dash compatible

thomasWos committed:
Fix typo

lessuselesss commented:
Hello, 

Thank you so much for the valuable feedback. I'm excited to contribute and am happy (and was hoping!!) to have someone from the nix community overseeing contributions here! 

Nice catch on finding the culprit, I'll investigate some workarounds 🙇

YoungPhlo committed:
docs: Update "What Did You Get Done This Week? 5" spaces notes

jzvikart committed:
Change CI trigger to 'pull_request'

odilitime committed:
Merge branch 'develop' into main

odilitime committed:
Merge pull request #1170 from thomasWos/main

fix: Fix typo in multiversx plugin prompt for creating token

odilitime committed:
Merge branch 'develop' into docs/community-streams

odilitime committed:
Merge pull request #1174 from YoungPhlo/docs/community-streams

docs: Update "What Did You Get Done This Week? 5" spaces notes

odilitime commented:
I don't like the hardcoded versions, maybe another dev can offer a better suggestions on how to get the latest version

like `git describe --tags --abbrev=0`

ai16z-demirix committed:
test: adding tests for runtime.ts. Modified README since we switched to vitest

Commit:
fixed CONTRIBUTING.md file Issue: 1048

madjin committed:
Merge branch 'develop' of github.com:ai16z/eliza into new_docs

mcp97 committed:
Merge pull request #1191 from ileana-pr/main

docs: fixed CONTRIBUTING.md file Issue: 1048

mcp97 committed:
Merge branch 'develop' into main

mcp97 committed:
Merge pull request #1190 from ai16z-demirix/main

test: adding tests for runtime.ts. Modified README since we switched to vitest

lessuselesss commented:
@odilitime 

Thanks for the feedback! I've updated the implementation to automatically detect versions from package.json instead of hardcoding them. The flake now:

- Reads Node.js version from package.json's engines field
- Extracts pnpm version from packageManager field
- Automatically handles version updates when package.json changes
- No manual version or hash updates needed

This should make it more maintainable and reduce the risk of version mismatches.

Commit:
feat(nix): add development environment flake

- Add Nix flake for reproducible development environment
- Auto-detect Node.js and pnpm versions from package.json
- Configure development tools (gcc, python3, vips, etc.)
- Add shell hook with helpful development commands

lessuselesss commented:
I've reset the branch to resolve the merge conflicts and provide a clean implementation. The core functionality remains the same:
- Auto-detecting versions from package.json
- Setting up the development environment
- Configuring all necessary tools

The previous discussion and feedback has been incorporated into this cleaner version.

HashWarlock commented:
> @odilitime
> 
> Thanks for the feedback! I've updated the implementation to automatically detect versions from package.json instead of hardcoding them. The flake now:
> 
> * Reads Node.js version from package.json's engines field
> * Extracts pnpm version from packageManager field
> * Automatically handles version updates when package.json changes
> * No manual version or hash updates needed
> 
> This should make it more maintainable and reduce the risk of version mismatches.
> 
> Alternatively,
> 
> @HashWarlock would an overlay for nodejs/pnpm be more appropriate nix'ing here?

Probably best to do an overlay bc whenever I run the current `nix develop` it will try to install package that is not present in nixpkgs. 
![telegram-cloud-photo-size-1-5122966996043017587-y](https://github.com/user-attachments/assets/e3f1ca5b-19e5-4e94-ae3c-6701cdd9452f)

How do the tests work in your nix environment?

Commit:
fix print alignment

tomguluson92 committed:
Update README_CN.md

Add more details in the CN version of README

9547 committed:
docs(cn): add python2.7

Signed-off-by: 9547 <29431502+9547@users.noreply.github.com>

9547 committed:
docs(cn): rm duplicated -

Signed-off-by: 9547 <29431502+9547@users.noreply.github.com>

jzvikart committed:
Merge remote-tracking branch 'ai16z/develop' into jure/integration-tests

v1xingyue committed:
gitpod cicd bug

Sometimes we can't fetch tags, so let's fetch tags before.

lessuselesss commented:
@HashWarlock 

I ended up installing eliza on a fresh VM as per the readme, and found similar results to the below (esp. #4)

Let me share what I've found from running the tests in the nix environment:

1. Core Tests Status:
- All 121 tests passed (with 1 skipped)
- Clean execution with no major issues
- Test duration: 1.52s

2. Plugin Tests Status:
- Several plugins executed tests successfully (e.g., solana plugin: 5 tests passed)
- Some failures were observed:
  - plugin-evm: 3 failed tests due to network connectivity issues with IoTeX endpoints
  - plugin-0g: Test process hangs and requires manual interruption
  - Several plugins have no test scripts configured yet

3. Regarding the nix environment:
- Tests are running through the nix shell environment successfully
- The test infrastructure itself works as expected
- We're seeing expected behavior with node modules and dependencies

4. Areas for Improvement:
- Some plugins need test script configurations
- Network-dependent tests (like EVM) might need mocking or better error handling
- The plugin-0g hanging issue needs investigation

Regarding your question about an overlay - yes, that sounds like a good approach so I've gone ahead and made an attempt.

Let me know if you'd like me to focus on any specific area or if you need more details about the test results.

Commit:
add overlays

shakkernerd committed:
Merge pull request #1177 from Sifchain/jure/integration-tests

feat: integration tests fixes + library improvements

marcNY committed:
Update README.md

Fix starter script: Add missing ‘cd’ command to navigate to ‘eliza-starter’ directory

odilitime committed:
Merge pull request #1201 from 9547/doc-cn

docs(cn): add python 3.7

odilitime committed:
Merge pull request #1207 from v1xingyue/v1xingyue-patch-1

fix: gitpod cicd bug

odilitime committed:
Merge branch 'develop' into patch-1

odilitime committed:
Merge pull request #1209 from marcNY/patch-1

docs: Update README.md

odilitime committed:
add missing change directories

odilitime committed:
Merge branch 'develop' into patch-1

odilitime committed:
Merge pull request #1196 from tomguluson92/patch-1

docs: Update "CN README" with more details

madjin committed:
add wip docs

madjin committed:
Merge branch 'develop' of github.com:ai16z/eliza into new_docs

madjin committed:
update api docs

mcp97 committed:
Merge pull request #1211 from ai16z/new_docs

cygaar committed:
Improve generation prompt

cygaar committed:
Update

odilitime committed:
postgres needs the user to exist before you can add a participant

cygaar committed:
Merge pull request #1217 from cygaar/improve_twitter

fix: improve twitter post generation prompt

tcm390 committed:
support image message

tcm390 committed:
check content type

tcm390 committed:
correct content type

odilitime committed:
Merge branch 'main' into postgres-fix

tcm390 committed:
clean code

tcm390 committed:
handle http image

Commit:
add visual flow to quickstart guide

Commit:
add visual flow to quickstart

mcp97 committed:
Merge pull request #1220 from ai16z/tcm-telegram-image

fix: Allow the bot to post messages with images generated by the imageGenerationPlugin on Telegram.

mcp97 committed:
Merge pull request #1219 from odilitime/postgres-fix

fix: postgres needs the user to exist before you can add a participant

lessuselesss committed:
Merge branch 'develop' into 1142--add-nix-flake-support

Commit:
add error handling

Commit:
add error handling

Commit:
remove debugging

Commit:
add flake.lock

lessuselesss commented:
I've installed NixOS and I'm seeing the same printout you were referring to when entering the nix dev environment.

For whatever reason nix-darwin wasn't having any issues.

Will try to figure out what is what here.

HashWarlock commented:
> on nix-darwin wasn't having any issues.
> 
> Will try to figure out what is what here.

Can you use nvm to install? This is how they recommend on node website now.
```
# installs nvm (Node Version Manager)
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.40.1/install.sh | bash

# download and install Node.js (you may need to restart the terminal)
nvm install 23

# verifies the right Node.js version is in the environment
node -v # should print `v23.4.0`

# verifies the right npm version is in the environment
npm -v # should print `10.9.2`
```

Commit:
add dns

Commit:
fix(nix): correct Node.js fetching in flake.nix

Commit:
refactor(nix): use nixpkgs nodejs instead of custom fetch

Commit:
remove networking/dns

Commit:
update shellHook print

Commit:
give add null hash

Commit:
fix(nix): make Node.js derivation platform-aware

- Remove Linux-specific ELF patching on macOS systems
- Conditionally include build inputs based on platform
- Ensure proper executable permissions for Node.js binaries

This fixes the development shell initialization on macOS while
maintaining Linux compatibility.

Commit:
fix fetching

Commit:
separate concerns

Commit:
simplify flake

Commit:
fix flake

lessuselesss commented:
@HashWarlock This approach simplified things greatly, cheers!

My NixOS vm isn't working, and won't install (apple hypervisor), so if anyone weigh in on how it's working that would be great.

HashWarlock commented:
> @HashWarlock This approach simplified things greatly, cheers!
> 
> My NixOS vm isn't working, and won't install (apple hypervisor), so if anyone weigh in on how it's working that would be great.

Testing now. Will respond back soon with results! Great work :)

lessuselesss commented:
@HashWarlock A bit disappointing, but I suppose those running NixOS as a daily driver are more equipped or ready to deal with such things.

Here's what I've tried thus far to get nodejs/pnpm not running off of nixpkgs, by referencing the package.json 

**Derivations** - No network access in the sandbox, which was getting us the initial errors you found. 

**Overlays** - Couldn't (yet) get it working/enter the development environment because of the dynamic logic. Moved it out of the flake.nix itself and made a solid attempt with nvfetcher (was aiming to have github actions perhaps update the .toml it reference with the url/hashes the flake would need. 

**NVM Flake** - I tried picking up where another project [nix-node](https://github.com/lessuselesss/nix-node) left off, to implement nvm-likeness. 

I'd like to get the overlays working but as per your point, it might be "fine enough" for now by addressing NixOS users in the README. 

If there are any next steps or other things to consider let me know, otherwise I'll maybe try to get my hands dirty a bit deeper into the project. Cheers for the support.

lalalune commented:
Gonna close this due to inactivity but we can revisit in v2
